lillisp.go

```diff
@@ -3,15 +3,78 @@ package main
 import (
 	"bufio"
 	"fmt"
+	"log"
 	"os"
+	"strconv"
 	"unicode/utf8"
 )
 
+// Atom ...
+type Atom interface{}
+
+// Pair represents the fundamental "cons-pair" in LISP-like languages
 type Pair struct {
 	car interface{}
 	cdr interface{}
 }
 
+var funcMap = map[string]func(Atom, Atom) Atom{
+	"+": func(a Atom, b Atom) Atom { return toInt(a) + toInt(b) },
+	"-": func(a Atom, b Atom) Atom { return toInt(a) - toInt(b) },
+	"*": func(a Atom, b Atom) Atom { return toInt(a) * toInt(b) },
+	"/": func(a Atom, b Atom) Atom { return toInt(a) / toInt(b) },
+}
+
+func toInt(a Atom) int {
+	if val, ok := a.(int); ok == true {
+		return val
+	} else if val, ok := a.(string); ok == true {
+		if intVal, err := strconv.Atoi(val); err == nil {
+			return intVal
+		}
+	}
+
+	log.Panicf("Could not convert %v to Number\n", a)
+	return 0 // technically will never reach here
+}
+
+func toFunc(p interface{}) (func(a Atom, b Atom) Atom, bool) {
+	op, ok := p.(string)
+	if !ok {
+		return nil, false
+	}
+	if funcMap[op] == nil {
+		return nil, false
+	}
+	return funcMap[op], true
+}
+
+func toPair(p interface{}) (*Pair, bool) {
```

Comment:
Doesn't this do exactly what `pair, ok := p.(*Pair)` does? Why bother with this function?

Reply:
Yeah, it does...  this function can be killed.